File: C/2.抽象工厂模式/src/ShapeFactory.c
/**
 * @file ShapeFactory.c
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2020-07-16
 *
 * @copyright Copyright (c) 2020
 *
 */

#include <stdlib.h>
#include <string.h>

#include "FactoryStruct.h"
#include "ShapeStruct.h"

extern struct Shape* rectangle_create(void);
extern struct Shape* square_create(void);
extern struct Shape* circle_create(void);

static struct Shape* factory_get_shape(struct Factory* factory, const char* shape_type) {
    if (NULL == shape_type) return NULL;

    if (0 == strcasecmp("CIRCLE", shape_type)) {
        return circle_create();
    } else if (0 == strcasecmp("RECTANGLE", shape_type)) {
        return rectangle_create();
    } else if (0 == strcasecmp("SQUARE", shape_type)) {
        return square_create();
    }

    return NULL;
}

static struct Color* factory_get_color(struct Factory* factory, const char* color_type) { return NULL; }

struct Factory* shape_factory_create(void) {
    struct Factory* factory = (struct Factory*)calloc(1, sizeof(struct Factory));
    if (NULL == factory) return NULL;

    factory->get_color = factory_get_color;
    factory->get_shape = factory_get_shape;

    return factory;
}

File: Objective-C/2.AbstractFactoryPattern/ShapeFactory.m
//
//  ShapeFactory.m
//  2.AbstractFactoryPattern
//
//  Created by Layne on 2020/10/18.
//

#import "ShapeFactory.h"
#import "Circle.h"
#import "Rectangle.h"
#import "Square.h"

@implementation ShapeFactory

- (id<ColorProtocol>)getColor:(NSString*)colorType {
    return nil;
}

- (id<ShapeProtocol>)getShape:(NSString*)shapeType {
    if (shapeType == nil) {
        return nil;
    }
    if ([shapeType isEqualToString:@"CIRCLE"]) {
        return [[Circle alloc] init];
    } else if ([shapeType isEqualToString:@"RECTANGLE"]) {
        return [[Rectangle alloc] init];
    } else if ([shapeType isEqualToString:@"SQUARE"]) {
        return [[Square alloc] init];
    }
    return nil;
}

@end
